src/operator/tensor/dot-inl.cuh

```diff
@@ -895,6 +988,131 @@ inline void DotCsrRspDnsImpl(const OpContext& ctx,
+          DType* csc_data_ptr = NULL;
+          unsigned long long* csc_indices_ptr = NULL;
+          unsigned long long* csc_indptr_ptr = NULL;
+          unsigned long long* col_counters = NULL;
+          size_t ull_mem_size = sizeof(unsigned long long);
+          void* temp_storage = NULL;
+          size_t temp_storage_bytes = 0;
+          CType out_num_rows = ret->shape()[0];
+          CType out_num_cols = ret->shape()[1];
+          // Get necessary temporary storage amount
+          cub::DeviceScan::ExclusiveSum(NULL,
+                                        temp_storage_bytes,
+                                        csc_indices_ptr,
+                                        csc_indices_ptr,
+                                        csr_cols+1,
+                                        Stream<gpu>::GetStream(s));
+          temp_storage_bytes += (ull_mem_size - (temp_storage_bytes % ull_mem_size));
+          Tensor<gpu, 1, char> workspace =
+            ctx.requested[0].get_space_typed<gpu, 1, char>(
+              Shape1(nnz*sizeof(DType) + nnz*ull_mem_size +
+                     2*(csr_cols + 1)*ull_mem_size +
+                     temp_storage_bytes),
+              s);
+          csc_indices_ptr = reinterpret_cast<unsigned long long*>(workspace.dptr_);
+          csc_indptr_ptr = reinterpret_cast<unsigned long long*>(
+                             workspace.dptr_ + nnz*ull_mem_size);
+          col_counters = reinterpret_cast<unsigned long long*>(
+                           workspace.dptr_ + nnz*ull_mem_size + (csr_cols+1)*ull_mem_size);
+          csc_data_ptr = reinterpret_cast<DType*>(workspace.dptr_ + nnz*ull_mem_size +
+                                                  2*(csr_cols+1)*ull_mem_size);
+          temp_storage = reinterpret_cast<void*>(workspace.dptr_ + nnz*sizeof(DType) +
+                                                 nnz*ull_mem_size + 2*(csr_cols+1)*ull_mem_size);
+          mxnet_op::Kernel<mxnet_op::set_zero, gpu>::Launch(
+            s, dns_rows*csr_cols, ret->data().dptr<DType>());
+          // Reset values for indptr, ready for histogramming
+          mxnet_op::Kernel<mxnet_op::set_zero, gpu>::Launch(
+            s, csr_cols + 1, csc_indptr_ptr);
+          // Histogramming on col id
+          mxnet_op::Kernel<CsrTransHistogramKernel, gpu>::Launch(
+            s, nnz, csr_indices.dptr<IType>(), csc_indptr_ptr, nnz);
+          cub::DeviceScan::ExclusiveSum(temp_storage,
+                                        temp_storage_bytes,
+                                        csc_indptr_ptr,
+                                        csc_indptr_ptr,
+                                        csr_cols+1,
+                                        Stream<gpu>::GetStream(s));
+          // Reset values for col_counter, ready for the final transform
+          mxnet_op::Kernel<mxnet_op::set_zero, gpu>::Launch(
+            s, csr_cols+1, col_counters);
```

Comment:
Can we rename "csr_cols" to "num_csr_cols"? It's not obvious from its name about whether it's a number, or a pointer to data, unless reading backward in code